### ISSUE TITLE
Unable to install in Linux Mint 18.3

### ISSUE THREAD

**Issue by vertigo220** [KengoSawa2/RapidCopy#3]
I've been using FastCopy in Windows for years now and love it, but have recently decided to make the switch to Linux (Mint Cinnamon 18.3). I was initially concerned about losing FastCopy's functionality, but the other day I found RapidCopy, and was very excited to get it installed. Unfortunately, I spent several hours that day trying unsuccessfully to get it working.

I first tried using Qt Creator as suggested by the README.md file (which doesn't give any more help beyond that), but when I load and debug or build the rapidcopy_main.pro file, I get the following error:

`/home/Downloads/RapidCopy-master/main.cpp:10: error: QLoggingCategory: No such file or directory`

After doing some digging and not being able to find a solution, I then tried running the following commands in a terminal:

```
$ qmake -project
$ qmake
$ make
```

Doing this initially gave me this error:

```
g++ -c -m64 -pipe -O2 -Wall -W -D_REENTRANT -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -DQT_SHARED -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4/QtCore -I/usr/include/qt4/QtGui -I/usr/include/qt4 -I. -I. -I. -o aboutdialog.o aboutdialog.cpp
In file included from ./mainwindow.h:16:0,
                 from aboutdialog.cpp:12:
./finactdialog.h:12:24: fatal error: QMediaPlayer: No such file or directory
compilation terminated.
Makefile:467: recipe for target 'aboutdialog.o' failed
make: *** [aboutdialog.o] Error 1
```

Up to this point, I had installed Qt Creator 3.5.1 and qt5-default (and possibly qt5-qmake, though I'm not sure, but it is installed now). I didn't download/install anything from the Qt website (which has changed, by the way, since the readme file was written) as their download server kept timing out, and I assumed I was getting what was needed through the package manager. I did a bunch of searching online and tried various things, installing and uninstalling packages and editing the .pro file, but as of right now I'm stuck with the following error when issuing the terminal commands above:

```
g++ -c -m64 -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -I. -I. -isystem /usr/include/x86_64-linux-gnu/qt5 -isystem /usr/include/x86_64-linux-gnu/qt5/QtGui -isystem /usr/include/x86_64-linux-gnu/qt5/QtCore -I. -I/usr/lib/x86_64-linux-gnu/qt5/mkspecs/linux-g++-64 -o aboutdialog.o aboutdialog.cpp
In file included from aboutdialog.cpp:9:0:
aboutdialog.h:11:19: fatal error: QDialog: No such file or directory
compilation terminated.
Makefile:598: recipe for target 'aboutdialog.o' failed
make: *** [aboutdialog.o] Error 1
```

There's no difference if I issue the commands as sudo. I cannot figure out how to solve this QDialog error, or why I'm getting different errors using the terminal than with Qt Creator, or how to solve the error in Qt Creator. Everywhere I look, I keep seeing the same suggestions, which don't work. These are to make sure the .pro file has various "QT +=" statements: gui/widgets/multimedia, which it does, and core, which I added without effect. I've tried putting them all on their own lines, also with no change. I ended up accessing the Qt download server directly and installing qt-opensource-linux-x64-5.10.0.run and QtInstallerFramework-linux-x64.run. I tried installing Qt Creator 4.5.0 as well as Qt-sdk. I've installed (or attempted to install, don't remember what worked and what didn't) libqt4-dev and libqt5-dev. I've run "sudo ldconfig" after installing the various packages, to make sure that's not the issue, but nothing. I've probably tried other stuff I haven't mentioned simply because I don't remember everything I did. I installed/uninstalled/reinstalled so much I broke the package manager to the point it could fix itself with the 'Fix Broken Packages' command or with the terminal commands, and I had to go through one by and and check and install dependencies. So I think it's safe to say I've tried everything within my capabilities, and so I'm hoping for some help. It would be really nice if someone could build a package for it, or at least provide detailed instructions on how to install it.

**Comment by KengoSawa2** [comment]
Hello.

Thank you for your detailed.
I understand your effort very well.

I have not tried building with Mint yet.
But for Mint users I should investigate the problem.
From now on I will create a situation close to your environment and investigate.
wait a minute.

**Comment by KengoSawa2** [comment]
I could :)

![mint18_3](https://user-images.githubusercontent.com/14103893/36135544-232052e6-10cf-11e8-99ca-d85d89300138.png)

Mint is similar to Ubuntu.
Although there were some problems, almost all of it was built like Ubuntu envrionment.
I will write the procedure, so please refer.

1. Install Qt 5.6.3 without source( I use qt-unified-linux-x64-3.0.2-online.run)
2. Install QtCreator from "sudo apt-get install qtcreator" (5.6.3 QtCreator has crash bug)
3. Install g++ from "sudo apt install g++"
4. Install libbsd from "sudo apt install libbsd-dev"
5. Install libxattr from "sudo apt install libattr1-dev"
6. Install libacl from "sudo apt install libacl1-dev"

7. QtCreator run with terminal command "qtcreator"
8. Open Project "rapidcopy_main.pro"
9. Choose Projects->Details->Manage->"Build&Run"->"Kits"->"Qt version"->"Manage"->select your "qmake" path
![qtkitselect](https://user-images.githubusercontent.com/14103893/36136068-f76f253e-10d1-11e8-85f8-2368547c5f60.png)

10. QtCreator Build and Run.

Cheers!

**Comment by vertigo220** [comment]
Thanks for the quick work on this; unfortunately, I'm still not able to get it working. I can't do step 1 because Qt's download server is horrible. It gets to 99%, sits there for several minutes or more, then times out. I can only seem to download stuff directly, but any online installers are useless. I already had qt-opensource-linux-x64-5.10.0.run (http://download.qt.io/official_releases/qt/5.10/5.10.0/) and, since it's a newer version, assumed it would work. I'm not sure if the issues I'm experiencing are due to not having the right version, having the complete wrong thing, or something else entirely. I also don't understand how qt-unified-linux-x64-3.0.2-online.run = 5.6.3 (since it says 3.0.2 in the file name) or what you mean by install it without source. I'm assuming you mean to uncheck source in the install options? And, assuming the 5.10.0 file I tried won't work for some reason, is there an offline installer that will?

When I do steps 3-6 it doesn't do anything, as all that stuff is already installed.

I couldn't quite follow your steps in #9, but I'm pretty sure I did what you meant. I didn't even get to the point of building the project (#10), though. I tried a debug first, and after counting 304 issues it popped up with an error stating "Could not find the executable, please specify one."

Finally, assuming I finally do get this working, can I remove Qt afterward? It's a massive program and if I don't need it, I don't want it taking up space. I know how to remove the Qt Creator with apt, but how would I uninstall the stuff from the .run file? Would I just delete the installed directory?

I've used Linux a bit here and there for years, but this is my first time seriously trying to use it, as I'm trying to switch over from Windows, and while I'm getting the hang of most things so far, I can't for the life of me figure out how to install programs. Almost every single one I've tried to install that isn't through the package manager or a PPA has been extremely difficult to impossible, and even some of the PPA installs haven't been easy. It's very frustrating and I fail to see why it's so hard. That's not a knock against you, but Linux as a whole, and if someone could explain to me why it is that way and hopefully point me in the right direction for figuring out how to make sense of it so I don't spend the next few months trying to install a half dozen programs, I'd really appreciate it. I just really don't want to have to stay with Windows.

**Comment by KengoSawa2** [comment]
Indeed, I understand your feelings well.
I am helping people of similar level in my company work.
I will understand your feelings well.

Let me solve your problem one by one as far as I can help.

You are stuck at # 9.

First of all, although I am a version of Qt, I recommend 5.6.3.
It seems to be able to download from the following, please try.
http://download.qt.io/official_releases/qt/5.6/5.6.3/
You can use "qt-opensource-linux-x64-5.6.3.run" in the above.

Then refer to the image below and make sure that the correct Qt library can be specified for QtCreator.
![qtadded](https://user-images.githubusercontent.com/14103893/36188312-a85c06b6-118e-11e8-817a-284eaff725d1.png)

Also, build directory
If it is correctly specified, the result should be output to "Build directory".
![builddir](https://user-images.githubusercontent.com/14103893/36188317-ae67ef02-118e-11e8-9406-485e51a4ad83.png)


If you do the same thing and you do not get it, QtCreator's
Please tell me the contents of "Compile Output".
![buildsucess](https://user-images.githubusercontent.com/14103893/36188334-c922ab02-118e-11e8-8f2c-7b9bcf8cea1f.png)


Since warnings should be huge, it would be better to zip the txt file by copying and pasting everything as text.

I will place an example of successful build, so please refer to it.
![buildsucess](https://user-images.githubusercontent.com/14103893/36188340-da0227f4-118e-11e8-98ff-00fb33441f9d.png)


About Qt library handling after building, Qt library file is necessary.
RapidCopy will not work if you delete everything.
You need to extract the necessary libraries and place them in the appropriate places,
I think that the data size was about 100 MB.
If you can tolerate it, I will teach you after a successful build.


The Linux world is the truth world of computers, the technology itself.
What you irritate and struggling is beautifully hidden in Windows and Mac.
Linux can not avoid this problem. I do not think you need to be an expert in computers, but I think you have to study somewhat about the concepts and basics of programs and compilation.

However, I also know that the current mechanism is complicated.
Because I want to make it easy to distribute it internally,
We are thinking a bit about distributing RapidCopyn by rpm command and apt-get command.
It takes time to execute. I do not even know the distribution mechanism at all.

**Comment by vertigo220** [comment]
It worked, but still with a bunch of issues. I first ran a debug, then a build (which I'm assuming basically do the same thing), and both had >300 issues, but resulted in a program that runs, though with all those issues, I don't know if it's built and will perform properly. I do see you have even more issues than me (which is strange, since it seems it should be the same) and don't seem bothered by it, so I'm guessing that's "normal," though I don't understand how or why. I've zipped and attached the log from both the debug and the build, though again I suspect they'll have the exact same info: [error logs.tar.gz](https://github.com/KengoSawa2/RapidCopy/files/1724279/error.logs.tar.gz).

Some other notes:

I'm still not clear on what components I have to install and which ones I can omit, but this is what I did:

![screenshot from 2018-02-14 08-37-27](https://user-images.githubusercontent.com/12024557/36208685-93e1a2f6-1191-11e8-9657-65ff8562c9ea.png)

Note it's taking up 0.25GB of space, which granted isn't a _lot_, and it certainly seems reasonable for a program like this, but it's quite a bit more than the ~100MB you were thinking of (unless you're referring to the parts that have to remain while the rest can be removed). Not a big deal, just figured it'd be nice to be able to regain that space if it was only needed to build the program. Really, at this point it's more the frustration with Qt and wanting it off my system. :P

I'm also confused about the versions. The file you had me download installed Qt Creator 4.0.3, which I was thinking was an old version because you referenced 5.6.3 QtCreator earlier, stating it has a crash bug, and told me to use apt to install it. So I did that again and the version that installed is 3.5.1. And I see now that 4.5.0 is the current version. So I guess the 5.6.3 is the sdk or framework and Qt Creator 4.0.3, the program, is a bit out of date but at least the same major version as the current one. I did use 4.0.3, with the 5.6.3 Qt, which seems like it could be a problem due to the crash bug you mentioned, but I haven't experienced any crashes, and it's the version shown in your screenshot. I don't know what version of Qt Creator you used, but since 3.5.1 (the one installed with apt) uses an older version of Qt and doesn't have any valid kits, I just used 4.0.3. Hopefully that all makes sense.

On a side note: I recently tested FastCopy against a few other programs and against Windows Explorer and was pleasantly surprised to find it was significantly faster than all of them. Of course, it takes longer overall than Windows does if verifying, since Windows doesn't do that, but it completes the copy much faster. And the extra time for verification is obviously expected and not at all an issue. I did note a couple things, though, which you may want to make more clear to users, are that it _appears_ to be slower than Windows, because Windows "cheats" and makes it look like it's done (the file transfer window closes) before it actually is, due to using write caching, whereas when FastCopy says it's done, it's actually done, which I prefer. But until I figured that out and disabled write caching on the test drives I was upset thinking FastCopy was actually significantly slower than Windows. Also, the option for whether or not to use the OS buffer doesn't explain itself very well and is therefore confusing. It says it will likely be slower, but in my testing I found it to be much, much faster, and so I enabled it and had it that way for a while, until the other day when I disabled it because transfers were going very slowly while I was running tests on a couple drives, which must have been tying up Windows' buffer. So in normal operation enabling it can vastly improve performance, which should be noted, but there should also be a warning about it. Anyway, it's an amazing program, which is why I'm trying so hard to get it working in Linux.

**Comment by KengoSawa2** [comment]
I looked log.
Build has been successful, congratulations!

About QtCreator version and crash problems, it is due to the difference between me and your environment.
I am using the virtualization environment, so it is prone to problems in drawing.
If there is no problem in your environment, you better not think too deeply.

The compression of disk space by the Qt library is somewhat annoying.
You seem to be having trouble consuming 0.25 GB, but unfortunately, I do not care much about the size difference.
Because in modern systems it does not have to worry about disk capacity so much.

I have something to apologize for.

About how to reduce the Qt library necessary to run RapidCopy,
My memory was wrong.
Since I was supposed to build and run, I have not tried the exact procedure to save disk space.

I think that disk space reduction is possible.
But it is difficult to you.
A problem will occur later. So I think that it is better to give up.

You seem to be terribly tired, but I do not have time to improve RapidCopy for everyone to use and make it easier to install.
In other words it means "it is based on knowledge for developers and programmers".

As another story,
I have a relationship with FastCopy developer, Mr Shirouzu, who is actually seeing various conversations.
It is close to the same Japanese people and the place where they live.
He is a genius and a hacker with amazing skills.
RapidCopy could not be made without his advice.
I am offering him development equipment.

If you are helped by FastCopy, please donate to Mr Shirouzu.

I might make improvements for distribution for people looking for RapidCopy for Linux.
It might be you who made that chance.
By getting a voice, we become motivated.
Your voice and irritated your voice this time, RapidCopy may be better.

Thank you for your effort.

**Comment by vertigo220** [comment]
No worries about the space used by Qt. I don't like to use so much space when not needed, but at the same time it's not like I'm hurting for space. No need to apologize about it.

I do fully intend to donate to FastCopy as well as some other projects, just now is not a good time for it. But it will happen.

Hopefully this thread will help others looking to make it work in Linux. If you want to make it even easier, though, let me know if I can help in any way, perhaps by testing or clarifying the instructions or whatever. Thanks so much for all the help!

**Comment by KengoSawa2** [comment]
I do not know if it will work with Mint,
We prepared a simple executable file that you can use easily, so please use it.
https://github.com/KengoSawa2/RapidCopy/tree/master/static_bin/ubuntu1604

**Comment by KengoSawa2** [comment]
Hi

Indeed, but I don't using Ubuntu.
Priority is a bit low, but I will try testing and building for Ubuntu 18 in the near future.
wait a moment please.
Even if it is late, I will try it in January next year:)

**Comment by KengoSawa2** [comment]
Add RapidCopy static binary for Ubuntu 18.04
Try it.

https://github.com/KengoSawa2/RapidCopy/tree/master/static_bin/ubuntu1804
